Packet++/header/IcmpV6Layer.h

```diff
@@ -0,0 +1,300 @@
+	while values in the range from 128 to 255 (high-order bit is 1) indicate an information message. */
+	uint8_t type;
+	/** The code field value depends on the message type and provides an additional level of message granularity */
+	uint8_t code;
+	/** The checksum field provides a minimal level of integrity verification for the ICMP message */
+	uint16_t checksum;
+};
+#pragma pack(pop)
+
+/**
+ * @struct icmpv6_echo_hdr
+ * ICMP echo request/reply message structure
+ */
+#pragma pack(push, 1)
+typedef struct icmpv6_echo_hdr : icmpv6hdr
+{
+	/** the echo request identifier */
+	uint16_t id;
+	/** the echo request sequence number */
+	uint16_t sequence;
+} icmpv6_echo_hdr;
+#pragma pack(pop)
+
+/**
+ * @struct icmpv6_echo_request
+ * ICMPv6 echo request/reply message structure
+ */
+typedef struct icmpv6_echo_request
+{
+	/** a pointer to the header data */
+	icmpv6_echo_hdr *header;
+	/** most echo requests/replies contain some payload data. This is the data length */
+	size_t dataLength;
+	/** most echo requests/replies contain some payload data. This is a pointer to this data */
+	uint8_t *data;
+} icmpv6_echo_request;
+
+/**
+ * @typedef icmpv6_echo_reply
+ * ICMPv6 echo reply message structure, same as icmpv6_echo_request
+ */
+typedef icmpv6_echo_request icmpv6_echo_reply;
+
+/**
+ * @class IcmpV6Layer
+ * Abstract base class for ICMPv6 protocol layers. Cannot be instantiated and contains common logic for derived classes.
+ */
+class IcmpV6Layer : public Layer
+{
+  public:
+	virtual ~IcmpV6Layer() {}
+
+	/**
+	 * A static method that determines the ICMPv6 protocol type of ICMPv6 layer raw data by looking at the
+	 * icmpv6hdr#type field
+	 * @param[in] data The pointer to the beginning of an ICMPv6 byte stream
+	 * @param[in] dataLen The length of the byte stream
+	 * @return The specific ProtocolType or UnknownProtocol if it can not be determined
+	 */
+	static ProtocolType getIcmpv6Version(uint8_t *data, size_t dataLen);
+
+	/**
+	 * @param[in] type Type to check
+	 * @return True if the layer if of the given type, false otherwise
+	 */
+	bool isMessageOfType(ICMPv6MessageType type) const { return getMessageType() == type; }
+
+	/**
+	 * @return Get the ICMPv6 Message Type
+	 */
+	ICMPv6MessageType getMessageType() const;
+
+	/**
+	 * @return Get the code header field
+	 */
+	uint8_t getCode() const;
+
+	/**
+	 * @return Get the checksum header field in host representation
+	 */
+	uint16_t getChecksum() const;
+
+	/**
+	 * Identifies next layer and tries to create it
+	 */
+	void parseNextLayer();
+
+	/**
+	 * @return The length of the ICMPv6 header
+	 */
+	size_t getHeaderLen() const;
+
+	/**
+	 * Calculate ICMPv6 checksum field
+	 */
+	void computeCalculateFields();
+
+	OsiModelLayer getOsiModelLayer() const { return OsiModelNetworkLayer; }
+
+  protected:
+	IcmpV6Layer() = default;
+
+	IcmpV6Layer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet)
+		: Layer(data, dataLen, prevLayer, packet) {}

```

Comment:
I think we can make this method public for parsing ICMPv6 messages we don't yet know how to parse